### PR TITLE
Correct Vanilla serialization of LocalDateTime

### LINKED ISSUE #4007: UseBlazor exception: Cannot convert value to NodaTime.LocalDateTime
When trying blazor on a customer canvas, I got an error in the console:

```
Error parsing analysis in Blazor
... lots of customer data here ...
caught exception: "Cannot convert value to NodaTime.LocalDateTime" "[]"
```

The dates in the handler were `2020-07-26T01:12:40Z` and `2020-07-23T04:04:01Z`. The functions in the handler were `Date::now`, `Date::>=`, `Date::fromSeconds` and `Date::subtract`.

The trace had no function_results (I think there's can't be dates in inputs).

### PR DESCRIPTION
**Pull request by StachuDotNet**
The one in the NuGet package isn't compatible with the custom converter used by the OCamlCompatible serializer

Resolves #4007.

### CODE REVIEW

File: fsharp-backend/src/Prelude/Prelude.fs [hunk 1]
```diff
@@ -904,6 +904,27 @@ module Json =
       override _.Write(writer : Utf8JsonWriter, _ : Password, _options) =
         writer.WriteStringValue("Redacted")
 
+    // Since we're getting this back from OCaml in DDates, we need to use the
+    // timezone even though there isn't one in the type
+    type LocalDateTimeConverter() =
```
Comment (StachuDotNet) [inline fsharp-backend/src/Prelude/Prelude.fs:909]: I just copied and modified the one from OCamlCompat.

**Comment by StachuDotNet** [comment]
note: my branch name is wrong/backwards, but the commit has the right phrasing.

File: fsharp-backend/src/Prelude/Prelude.fs (hunk 1 above)
Comment (StachuDotNet) [reply fsharp-backend/src/Prelude/Prelude.fs:909]: thought about re-using some code, but decided against it (rule of 3)

Comment (pbiggar) [reply fsharp-backend/src/Prelude/Prelude.fs:909]: We'll delete the old one at some point I hope.

Comment (StachuDotNet) [reply fsharp-backend/src/Prelude/Prelude.fs:909]: exactly

**Comment by pbiggar** [comment]
I get nervous about changing these. Does any part of the backend serialize Dvals using this?

**Comment by StachuDotNet** [comment]
> I get nervous about changing these. Does any part of the backend serialize Dvals using this?

I don't know how to quickly check this.
Looking for `Vanilla` across `fsharp-backend/src`, I get this:

```
29 results - 14 files

fsharp-backend/src/ApiServer/Functions.fs:
  90    |> List.sortBy (fun fn -> fn.name)
  91:   |> Json.Vanilla.prettySerialize
  92  

fsharp-backend/src/ApiServer/Http.fs:
  107        use t = startTimer "serialize-json" ctx
  108:       addTag "json_flavor" "vanilla"
  109        ctx.Response.ContentType <- "application/json; charset=utf-8"

fsharp-backend/src/ApiServer/Json.fs:
   5  // client, but uses System.Text.Json. The intent is to be significantly faster than
   6: // old ocamlCompatible one. It doesn't differ from Prelude.Json.Vanilla very much,
   7  // and I suspect it could be

  66    options.Converters.Add(NodaConverters.InstantConverter)
  67:   options.Converters.Add(Prelude.Json.Vanilla.TLIDConverter())
  68:   options.Converters.Add(Prelude.Json.Vanilla.PasswordConverter())
  69:   options.Converters.Add(Prelude.Json.Vanilla.RawBytesConverter())
  70    options.Converters.Add(fsharpConverter)

fsharp-backend/src/ApiServer/Ui.fs:
  59    |> File.readfile Config.Webroot
  60:   |> Json.Vanilla.deserialize<Map<string, string>>
  61    |> Map.remove "__date"

  69  let prodHashReplacementsString : string =
  70:   prodHashReplacements |> Json.Vanilla.serialize
  71  

fsharp-backend/src/LibBackend/Analytics.fs:
  40          heapioMetadataJson
  41:         |> Option.map Json.Vanilla.deserialize<Map<string, string>>
  42          |> Option.defaultValue Map.empty

fsharp-backend/src/LibBackend/Canvas.fs:
  315      | Some AllOrigins -> Sql.jsonb "\"*\""
  316:     | Some (Origins ss) -> ss |> Json.Vanilla.serialize |> Sql.jsonb
  317    Sql.query

  566            | PT.Toplevel.TLHandler ({ pos = pos })
  567:           | PT.Toplevel.TLDB { pos = pos } -> Some(Json.Vanilla.serialize pos)
  568            | PT.Toplevel.TLType _ -> None

  639      json
  640:     |> Json.Vanilla.deserialize<OT.oplist<OT.RuntimeT.fluidExpr>>
  641      |> OT.Convert.ocamlOplist2PT

fsharp-backend/src/LibBackend/EventQueueV2.fs:
  325            |> UTF8.ofBytesUnsafe
  326:           |> Json.Vanilla.deserialize<NotificationData>
  327          let deliveryAttempt =

  344            { id = id; canvasID = canvasID }
  345:           |> Json.Vanilla.serialize
  346            |> Google.Protobuf.ByteString.CopyFromUtf8

fsharp-backend/src/LibBackend/Init.fs:
  59  
  60:     Json.Vanilla.registerConverter (
  61        QueueSchedulingRules.WorkerStates.STJJsonConverter.WorkerStateConverter()

fsharp-backend/src/LibBackend/Pusher.fs:
  61    : unit =
  62:   push canvasID "new_trace" (Json.Vanilla.serialize (traceID, tlids))
  63  

  65  let pushNew404 (canvasID : CanvasID) (f404 : TraceInputs.F404) =
  66:   push canvasID "new_404" (Json.Vanilla.serialize f404)
  67  

  69  let pushNewStaticDeploy (canvasID : CanvasID) (asset : StaticAssets.StaticDeploy) =
  70:   push canvasID "new_static_deploy" (Json.Vanilla.serialize asset)
  71  

  80    : unit =
  81:   push canvasID "worker_state" (Json.Vanilla.serialize ws)
  82  

fsharp-backend/src/LibBackend/Session.fs:
  38      let date = read.instant "expire_date"
  39:     let data = Json.Vanilla.deserialize<JsonData> serializedData
  40  

  61      let sessionData =
  62:       Json.Vanilla.serialize { username = string username; csrf_token = csrfToken }
  63  

fsharp-backend/src/LibExecution/DvalReprInternalNew.fs:
  127  let toRoundtrippableJsonV0 (dv : RT.Dval) : string =
  128:   dv |> RoundtrippableSerializationFormatV0.fromRT |> Json.Vanilla.serialize
  129  

  131    json
  132:   |> Json.Vanilla.deserialize<RoundtrippableSerializationFormatV0.Dval>
  133    |> RoundtrippableSerializationFormatV0.toRT

fsharp-backend/src/LibService/Rollbar.fs:
  179  
  180:   let queryStr = Json.Vanilla.serialize query
  181    let dataset = Config.honeycombDataset

fsharp-backend/src/Prelude/Prelude.fs:
  878  
  879:   module Vanilla =
  880  

  989    module OCamlCompatible =
  990:     // CLEANUP: get rid of OCamlCompatible serializers, replacing it with Vanilla.
  991      // AFAIK, the only places we have to use it is for actual ocamlinterop, and all

fsharp-backend/src/Wasm/Wasm.fs:
  311          try
  312:           Ok(Json.Vanilla.deserialize<ClientInterop.performAnalysisParams> message)
  313          with

  338      }
  339:     |> Task.map Json.Vanilla.serialize
  340      |> Task.map EvalWorker.postMessage

  394    ]"""
  395:   |> Json.Vanilla.deserialize<ClientInterop.performAnalysisParams>
```

Is that helpful?

**Comment by pbiggar** [comment]
Some of the pusher ones look like they might cause problems, but the downside to them is low - we'll get alerted by rollbar and we can fix them then.

However, fsharp-backend/src/LibExecution/DvalReprInternalNew.fs could be a real problem. It's fair to expect a few values will be saved.

Can you add a test that the old format continues to be correctly deserialized by the new deserialier?

**Comment by StachuDotNet** [comment]
> Can you add a test that the old format continues to be correctly deserialized by the new deserialier?

To be sure - what "old format" are you referring to - LocalDateTimes serialized by the NuGet-powered converter, somehow ending up in a trace or something?

**Comment by pbiggar** [comment]
> > Can you add a test that the old format continues to be correctly deserialized by the new deserialier?
> 
> To be sure - what "old format" are you referring to - LocalDateTimes serialized by the NuGet-powered converter, somehow ending up in a trace or something?

Before this PR, `DDate`s were serialized using the old version of this formatter (that is, the code that you are replacing here). Given that that serialized data presumably exists somewhere (the queues) we need to ensure that it can continue to be read.

**Comment by StachuDotNet** [comment]
Makes sense. I added a test - we seem to be OK?

I got the value by briefly switching the serializer back, adding the test to quickly serialize some value (so I have the old format handy), reverting the serialization test, then adjusting the test to this.

File: fsharp-backend/tests/Tests/DvalRepr.Tests.fs [hunk 2]
```diff
@@ -107,6 +107,18 @@ let testDateMigrationHasCorrectFormats =
...
+// have since adjusted `Vanilla` to use a custom converter, and this test is to
+// ensure values serialized during the time where the NuGet package's converter
+// are able to be deserialized. The value here
+let testPreviousDateSerializionCompatibility =
+  test "previous date serialization compatible" {
+    let expected = RT.DDate(NodaTime.Instant.UnixEpoch.toUtcLocalTimeZone ())
+    let actual =
+      Json.Vanilla.deserialize<RT.Dval> """["DDate","1970-01-01T00:00:00"]"""
+    Expect.equal expected actual "not deserializing correctly"
+  }
```
Comment (StachuDotNet) [inline fsharp-backend/tests/Tests/DvalRepr.Tests.fs:121]: I wasn't feeling very creative about naming this test! feels a bit meh, though maybe I'm nitpicky at myself here.

File: fsharp-backend/tests/Tests/DvalRepr.Tests.fs [hunk 3]
```diff
@@ -107,6 +107,18 @@ let testDateMigrationHasCorrectFormats =
...
+// We used a System.Text.Json converter supplied by a NuGet package for a bit,
+// but found that it was incompatible with the OCamlCompatible serializer. We
+// have since adjusted `Vanilla` to use a custom converter, and this test is to
+// ensure values serialized during the time where the NuGet package's converter
+// are able to be deserialized. The value here
+let testPreviousDateSerializionCompatibility =
+  test "previous date serialization compatible" {
+    let expected = RT.DDate(NodaTime.Instant.UnixEpoch.toUtcLocalTimeZone ())
+    let actual =
+      Json.Vanilla.deserialize<RT.Dval> """["DDate","1970-01-01T00:00:00"]"""
```
[the remark this answers is not in the document]
Comment (StachuDotNet) [reply fsharp-backend/tests/Tests/DvalRepr.Tests.fs:119]: :shrug: Maybe I did something wrong, but that's not what I found. I switched back the serializers to get this sample.

File: fsharp-backend/tests/Tests/DvalRepr.Tests.fs (hunk 2 above)
Comment (StachuDotNet) [reply fsharp-backend/tests/Tests/DvalRepr.Tests.fs:121]: better than `DvalReprInternal'`

File: fsharp-backend/tests/Tests/DvalRepr.Tests.fs (hunk 3 above)
[the remark this answers is not in the document]
Comment (StachuDotNet) [reply fsharp-backend/tests/Tests/DvalRepr.Tests.fs:119]: Maybe you can find a value in prod? Not sure if there's a trivial way to search for such.

**Pull request merged**